internal/langserver/handlers/did_change_watched_files.go

```diff
@@ -0,0 +1,181 @@
+package handlers
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/creachadair/jrpc2"
+	"github.com/hashicorp/terraform-ls/internal/document"
+	"github.com/hashicorp/terraform-ls/internal/job"
+	"github.com/hashicorp/terraform-ls/internal/protocol"
+	lsp "github.com/hashicorp/terraform-ls/internal/protocol"
+	"github.com/hashicorp/terraform-ls/internal/terraform/ast"
+	"github.com/hashicorp/terraform-ls/internal/uri"
+)
+
+func (svc *service) DidChangeWatchedFiles(ctx context.Context, params lsp.DidChangeWatchedFilesParams) error {
+	var ids job.IDs
+
+	for _, change := range params.Changes {
+		rawURI := string(change.URI)
+
+		rawPath, err := uri.PathFromURI(rawURI)
+		if err != nil {
+			svc.logger.Printf("error parsing %q: %s", rawURI, err)
+			continue
+		}
+
+		if change.Type == protocol.Deleted {
+			// We don't know whether file or dir is being deleted
+			// 1st we just blindly try to look it up as a directory
+			_, err = svc.modStore.ModuleByPath(rawPath)
+			if err == nil {
+				svc.removeIndexedModule(ctx, rawURI)
+				continue
+			}
+
+			// 2nd we try again assuming it is a file
+			parentDir := filepath.Dir(rawPath)
+			_, err = svc.modStore.ModuleByPath(parentDir)
+			if err != nil {
+				svc.logger.Printf("error finding module (%q deleted): %s", parentDir, err)
+				continue
+			}
+
+			// and check the parent directory still exists
+			fi, err := os.Stat(parentDir)
+			if err != nil {
+				if os.IsNotExist(err) {
+					// if not, we remove the indexed module
```

Comment:
When can this happen? The user deleted the only file and then subsequently the directory, too? If yes, then those would also be different events here and that's probably why you're just ensuring the directory is still present or not.

Reply:
This accounts for the deletion of a whole directory.

> If yes, then those would also be different events here

Frankly, I don't know, since LSP doesn't go into detail about how clients should or shouldn't group file events. In fact when I tried to `rm -rf submodule` in VS Code where submodule had some `*.tf` files, no events were fired at all - which may be a bug in the VS Code language client - unsure.

So in the face of uncertainty I'm just being a little more defensive here.

Reply:
Got it, thanks for the explanation and yes, it's still a little nebulous when it comes to understanding client-side event emits.

> may be a bug in the VS Code language client
Do you mean https://github.com/Microsoft/vscode-languageserver-node?

Reply:
It could be either VS Code itself or the language client you linked - it would take some debugging to find out.